Poll real health checks

2uasimojo opened the pull request:
With this PR, we poll [real osde2e health checks](https://github.com/openshift/osde2e/blob/041355675304a7aa371b7fbeea313001036feb75/pkg/common/cluster/clusterutil.go#L211) at one-minute intervals, forever. If they fail, we make sure there's an active silence. If they succeed, we make sure the silence is removed and we exit.

2uasimojo committed:
Add Makefile, update .gitignore

2uasimojo committed:
Pin k8s modules to v0.19.0

...per https://github.com/kubernetes/client-go/issues/741#issuecomment-737117098
to get around build errors

2uasimojo committed:
Poll real health checks

With this commit, we poll
[real osde2e health checks](https://github.com/openshift/osde2e/blob/041355675304a7aa371b7fbeea313001036feb75/pkg/common/cluster/clusterutil.go#L211)
at one-minute intervals, forever. If they fail, we make sure there's an
active silence. If they succeed, we make sure the silence is removed and
we exit.

2uasimojo commented:
/hold

Prereq https://github.com/openshift/osde2e/pull/681

2uasimojo commented on go.mod:
```diff
@@ -3,13 +3,60 @@ module github.com/iamkirkbater/osd-readiness-spike
+	github.com/go-delve/delve v1.5.1 // indirect
+	github.com/mattn/go-colorable v0.1.8 // indirect
+	github.com/mattn/go-runewidth v0.0.10 // indirect
+	github.com/openshift/osde2e v0.0.0-20210127203025-7b09b7d9f79d
```
This will need to be updated with a new hash once https://github.com/openshift/osde2e/pull/681 lands.

2uasimojo replied:
Done.

iamkirkbater commented on main.go:
```diff
@@ -46,61 +40,43 @@ type silenceResponse struct {
 }
 
 func main() {
-	clusterInit, err := getClusterCreationTime()
```
Is there a reason why we're getting rid of this cluster timestamp check?  I'm not sure off the top of my head how MCC pushes out changes to this job, so we'd want to confirm that it's not going to re-run the job on existing clusters when we push an update.

If it doesn't re-run the job on update, then we're safe to remove this, however if it does re-run the job this could silence an already-ready production cluster if it's experiencing a blip of something updating.

2uasimojo replied:
Yeah, I struggled with that. You're probably right that we should keep this check in place -- but from what @jharrington22 was telling me yesterday, one hour may be way too short. James, what do you think?

iamkirkbater replied:
The job _should_ be deployed within one hour via MCC, but really I'd be comfortable putting this to even the first 4 hours or so that a cluster is deployed, that would make testing easier and a bit more flexible.

2uasimojo replied:
Roger wilco, I'll do that. Mebbe I'll add support for an env var so we can configure it (including for testing).

2uasimojo replied:
This is done via 060f06b.

iamkirkbater commented on main.go:
```diff
@@ -46,61 +40,43 @@ type silenceResponse struct {
-	}
-	if silenceID == "" {
-		silenceID, err = createSilence()
+	for {
```
In osde2e they've found that the cluster can become ready but somewhere like 5-10 minutes later it just starts upgrading, so they've created a loop where they check to see that it's healthy 20 times in a row before continuing.  Should we implement something similar here?  CC: @jeefy for additional context

jeefy replied:
:100: -- We've often seen that a cluster will be "healthy" for a period of time then start updating operators. 

In our OSDe2e job runs we require 20 "Clean" healthchecks *_in a row_* before moving forward. 

That number was arbitrary, but I've even recently seen a cluster having 12 clean runs then going back to 0.

2uasimojo replied:
> 20 "Clean" healthchecks __in a row__

over what time period?

2uasimojo replied:
...looks like [30s](https://github.com/openshift/osde2e/blob/7e201f679bbe1a78b67701df77ebb69bcf768957/pkg/common/cluster/clusterutil.go#L102) x [20 by default](https://github.com/openshift/osde2e/blob/7e201f679bbe1a78b67701df77ebb69bcf768957/pkg/common/config/config.go#L540).

Which means that we're going to continue to silence alerts for a full 10 minutes after the cluster is healthy.

iamkirkbater replied:
I think that's fair.  At this point in the cluster lifecycle the cluster shouldn't really have much if any customer workload on it that this should disturb.

iamkirkbater replied:
@jeefy yesterday when I was working on the CI Watcher stuff I was looking at a cluster that had never became "healthy" after an upgrade and it was getting to 15 and even 18 before something would roll again and reset the clock.

jeefy replied:
Yep.... :fire:

2uasimojo replied:
Still working on this.

2uasimojo replied:
Done via 27771f4.

2uasimojo committed:
Move silence functions

Move silence-related functions into their own package for modularity.

2uasimojo committed:
Make max cluster age configurable

With this commit, the maximum cluster age (after which we will clear any
silences and succeed, regardless of health checks):
- Is configurable by setting `MAX_CLUSTER_AGE_MINUTES` in the container
environment.
- Defaults to two hours in the code.
- Is overridden to four hours in the Job manifest due to
[BZ #1921413](https://bugzilla.redhat.com/show_bug.cgi?id=1921413)

2uasimojo commented:
> /hold
> 
> Prereq [openshift/osde2e#681](https://github.com/openshift/osde2e/pull/681)

Prereq merged.

But still /hold (not that it's doing anything) for [multiple successful health checks](https://github.com/iamkirkbater/osd-cluster-ready-job/pull/1#discussion_r568616600)

2uasimojo committed:
Require multiple consecutive successful health checks

...before declaring the cluster truly healthy, removing silences, and
exiting.

By default, we require 20 consecutive successful health checks at
30-second intervals. These are tunable via container environment
variables.

We also add a tunable for the time to sleep after a failed health check,
which defaults to 60s (one minute).

2uasimojo commented:
> But still /hold (not that it's doing anything) for [multiple successful health checks](https://github.com/iamkirkbater/osd-cluster-ready-job/pull/1#discussion_r568616600)

This is done.

/hold cancel

(FWIW)

iamkirkbater reviewed (approved):
LGTM.

I know @jharrington22 said he wanted to review, so once he says this is good go ahead and merge it and build and push as you would.

jharrington22 commented on README.md:
```diff
@@ -1,22 +1,80 @@
+- [TO DO](#to-do)
 
-The silence takes effect for 1 hour, which allows any clusters that are having issues to alert an SRE for them to actually look at.
+This job silences alerts while Day2 configuration is loaded onto a cluster at initial provisioning, allowing it to not page on-call SREs for normal operations within the cluster.
```
Non blocker but its not only the day 2 configuration keeping these clusters from being ready. ClusterOperators is still observing operators from being ready as they transition through different phases while configuration from each is loaded.

jharrington22 commented on README.md:
```diff
@@ -1,22 +1,80 @@
+
+The silence initially takes effect for 1 hour.
+
+We poll cluster health using [osde2e health checks](https://github.com/openshift/osde2e/blob/041355675304a7aa371b7fbeea313001036feb75/pkg/common/cluster/clusterutil.go#L211)
+once a minute (this is [configurable](#failed_check_interval_seconds)),
+until they all report healthy 20 times in a row ([configurable](#clean_check_runs))
+on 30s intervals ([configurable](#clean_check_interval_seconds)).
+By default, we will clear any active silence and exit successfully if the cluster is (or becomes) more than two hours old ([configurable](#max_cluster_age_minutes)).
```
:+1:

jharrington22 reviewed (approved):
/lgtm

2uasimojo commented:
@iamkirkbater I don’t think I can merge this. Would you please do the honors?
